modules/fedimint-ln-client/src/lib.rs

```diff
@@ -1185,55 +1128,97 @@ impl LightningClientModule {
-                                Ok(_) => {}
-                                Err(_) => {
-                                    yield LnPayState::UnexpectedError { error_message: "Error occurred while waiting for the primary module's output".to_string() };
-                                    return;
+                            match client_ctx.await_primary_module_outputs(operation_id, change.clone()).await {
+                                Ok(_) => {
+                                    yield LnPayState::Success { preimage };
+                                }
+                                Err(e) => {
+                                    yield LnPayState::UnexpectedError { error_message: format!("Error occurred while waiting for the change: {e:?}") };
                                 }
                             }
+                        } else {
+                            yield LnPayState::Success { preimage };
                         }
-
-                        yield LnPayState::Success {preimage};
-                        return;
                     }
-                    Err(PayError::Refundable(block_height, error)) => {
-                        yield LnPayState::WaitingForRefund{ block_height, gateway_error: error.clone() };
-
-                        if let Ok(out_points) = self_ref.await_refund(operation_id).await {
-                            // need to await primary module to get refund
-                            if client_ctx.await_primary_module_outputs(operation_id, out_points).await.is_ok() {
-                                yield LnPayState::Refunded { gateway_error: error };
-                                return;
+                    Some(LightningPayStates::Refund(refund)) => {
+                        yield LnPayState::WaitingForRefund {
+                            error_reason: refund.error_reason.clone(),
+                        };
+
+                        match client_ctx.await_primary_module_outputs(operation_id, refund.out_points).await {
+                            Ok(_) => {
+                                yield LnPayState::Refunded {
+                                    error_reason: refund.error_reason,
+                                };
+                            }
+                            Err(e) => {
+                                yield LnPayState::UnexpectedError {
+                                    error_message: format!("Error occurred trying to get refund. Refund was not successful: {e:?}"),
+                                };
                             }
                         }
                     }
-                    _ => {}
+                    Some(state) => {
+                        yield LnPayState::UnexpectedError { error_message: format!("Found unexpected state during lightning payment: {state:?}") };
+                    }
+                    None => {
+                        error!("Unexpected end of lightning pay state machine");
```

Comment:
Should this panic or return an `UnexpetedError`? Otherwise we'll save the last returned state as the final outcome.

Reply:
I think it should return `UnexpectedError`

Reply:
Fixed